feat : Add authentication , Initial APIs and project initialization files

SkySingh04 opened the pull request:
# APIs Added

## Experiment Management APIs
- **SaveExperiment API**
  - **Implemented in** `pkg/apis/experiment/experiment.go` as `SaveExperiment`
  - Saves an experiment with provided request data and project ID
  - Tested in `experiment_test.go` using authenticated client

- **RunExperiment API**
  - **Implemented in** `pkg/apis/experiment/experiment.go` as `RunExperiment`
  - Executes a saved experiment using project ID and experiment ID
  - Tested in `experiment_test.go` using authenticated client

- **CreateExperiment API**
  - **Implemented in** `pkg/apis/experiment/experiment.go` as `CreateExperiment`
  - Creates and immediately runs an experiment
  - Tested in `experiment_test.go` using authenticated client

- **GetExperimentList API**
  - **Implemented in** `pkg/apis/experiment/experiment.go` as `GetExperimentList`
  - Retrieves a paginated list of experiments for a project
  - Tested in `experiment_test.go` using authenticated client

- **GetExperimentRunsList API**
  - **Implemented in** `pkg/apis/experiment/experiment.go` as `GetExperimentRunsList`
  - Retrieves a paginated list of experiment runs
  - Tested in `experiment_test.go` using authenticated client

- **DeleteChaosExperiment API**
  - **Implemented in** `pkg/apis/experiment/experiment.go` as `DeleteChaosExperiment`
  - Deletes an experiment by ID from a project
  - Tested in `experiment_test.go` using authenticated client

## Project Management APIs
- **ListProject API**
  - **Implemented in** `pkg/apis/project.go` as `ListProject`
  - Retrieves a list of projects using the provided credentials
  - Used in the client implementation

- **CreateProject API**
  - **Implemented in** `pkg/apis/project.go` as `CreateProjectRequest`
  - Creates a new project with the given name and credentials
  - Used in the client implementation

- **GetProjectDetails API**
  - **Implemented in** `pkg/apis/project.go` as `GetProjectDetails`
  - Fetches detailed information about projects
  - Used in the client implementation

## Authentication and Request Handling
- **Auth API**
  - **Implemented in** `pkg/apis/auth.go` as `Auth`
  - Authenticates the user with username and password
  - Returns an access token for API requests
  - Used by client for authenticated requests

- **SendRequest API**
  - **Implemented in** `pkg/apis/request.go` as `SendRequest`
  - Sends HTTP requests with authentication headers
  - Core utility used by all API functions

## Testing Approach
Our testing approach involves:
1. **Client-based testing** - Tests utilize the `NewLitmusClient` that authenticates with username/password
2. **Integration testing** - Tests make actual API calls to a running Litmus server
3. **Response validation** - Tests verify response structures, field values, and error handling
4. **Response marshaling tests** - Additional tests verify JSON unmarshaling for different response types

### "mongo: no documents in result" Errors
These tests fail because they can't find expected documents in the database:

1. **Experiment API**:
   - `TestSaveExperiment/successful_experiment_save` - Cannot find infrastructure details
   - `TestRunExperiment/successful_experiment_run` - Cannot find experiment to run
   - `TestDeleteChaosExperiment/successful_experiment_deletion` - Cannot find experiment to delete
   - `TestCreateExperiment/successful_experiment_creation_and_run` - Cannot find infrastructure details

This is because in our tests we were testing the real GraphQL API calls with a real mongodb database. This will be fixed in future iterations.

SkySingh04 committed:
Add LICENSE file , README and initialize Go module

Signed-off-by: Sky Singh <akashsingh2210670@gmail.com>

SkySingh04 committed:
Implement authentication and request handling APIs

Signed-off-by: Sky Singh <akashsingh2210670@gmail.com>

SkySingh04 committed:
Add main application logic and utility functions for Litmus client

Signed-off-by: Sky Singh <akashsingh2210670@gmail.com>

SkySingh04 committed:
Add Litmus client unit tests and update project API field name

Signed-off-by: Sky Singh <akashsingh2210670@gmail.com>

SkySingh04 committed:
Add GraphQL API integration for managing Chaos Experiments

- Introduced new package `experiment` with functions to create, run, list, and delete Chaos Experiments.
- Added GraphQL queries and types for handling experiment data.
- Updated `go.mod` and `go.sum` to include new dependencies for the GraphQL server.

This commit enhances the functionality of the Litmus client by enabling users to manage Chaos Experiments through GraphQL APIs.

Signed-off-by: Sky Singh <akashsingh2210670@gmail.com>

SkySingh04 committed:
Add unit tests for Chaos Experiment management

- Created a new test file `experiment_test.go` to implement unit tests for the Chaos Experiment functionalities.
- Added tests for saving, running, listing, and deleting experiments, ensuring proper response structures and validation.
- Included tests for JSON unmarshaling of response structures to verify data integrity.

This commit enhances the test coverage for the Chaos Experiment management features in the Litmus client.

Signed-off-by: Sky Singh <akashsingh2210670@gmail.com>

SkySingh04 committed:
Update Litmus client test configurations and project API field name

Signed-off-by: Sky Singh <akashsingh2210670@gmail.com>

SkySingh04 committed:
Refactor experiment tests to use LitmusClient for authentication

- Replaced the `setupTestCredentials` function with `setupTestClient` to create a LitmusClient instance for authentication in the experiment tests.
- Updated all relevant test functions to utilize the new client for API calls, ensuring consistent authentication handling across tests.

This change improves the maintainability and clarity of the test code by centralizing client setup.

Signed-off-by: Sky Singh <akashsingh2210670@gmail.com>

ispeakc0de commented on pkg/utils/constants.go:
```diff
@@ -0,0 +1,70 @@
+/*
+Copyright © 2021 The LitmusChaos Authors
```
```suggestion
Copyright © 2025 The LitmusChaos Authors
```

ispeakc0de commented on client_test.go:
```diff
@@ -0,0 +1,33 @@
+	assert.NoError(t, err)
+
+	projectID := uuid.New().String()
+	project, err := client.CreateProject("test-project-" + projectID)
```
```suggestion
	project, err := client.CreateProject(fmt.Sprintf("test-project-%s" ,projectID))
```

ispeakc0de commented on main.go:
```diff
@@ -0,0 +1,91 @@
+
+func main() {
+	// Get credentials from environment variables or flags
+	endpoint := getEnv("LITMUS_ENDPOINT", "http://localhost:8080")
```
Can we encapsulate all these fields within a Configuration struct and load the data at once?? 
refer: [struct](https://github.com/litmuschaos/litmus/blob/master/chaoscenter/graphql/server/utils/variables.go#L7) and [load](https://github.com/litmuschaos/litmus/blob/master/chaoscenter/graphql/server/server.go#L44)

SkySingh04 replied:
This has been added!

ispeakc0de commented on main.go:
```diff
@@ -0,0 +1,91 @@
+	if err != nil {
+		log.Fatalf("Failed to list projects: %v", err)
+	}
+	fmt.Printf("Projects: %+v\n", projects)
```
use logrus for logging by creating wrapper similar to https://github.com/litmuschaos/litmus-go/blob/master/pkg/log/log.go

SkySingh04 replied:
The logging for the entire codebases has been revamped to use logrus!

ispeakc0de commented on pkg/apis/auth.go:
```diff
@@ -0,0 +1,65 @@
+	Password string `json:"password"`
+}
+
+func Auth(input types.AuthInput) (types.AuthResponse, error) {
```
Define an interface and a corresponding struct, then list all the functions within the interface.

ispeakc0de replied:
same for all apis

SkySingh04 replied:
Could you share an existing example of this if present? Would prefer to follow best practices / implementation

ispeakc0de commented on pkg/apis/auth.go:
```diff
@@ -0,0 +1,65 @@
+	}
+
+	// Sending token as empty because auth server doesn't need Authorization token to validate.
+	resp, err := SendRequest(SendRequestParams{input.Endpoint + utils.AuthAPIPath + "/login", ""}, payloadBytes, string(types.Post))
```
use fmt.Sprintf for string formating

SkySingh04 replied:
refactored!

SkySingh04 committed:
Update constants.go

Co-authored-by: Shubham Chaudhary <ashubham314@gmail.com>
Signed-off-by: Sky Singh <akashsingh2210670@gmail.com>

ispeakc0de commented on pkg/apis/experiment/experiment.go:
```diff
@@ -0,0 +1,348 @@
+			return RunExperimentResponse{}, errors.New(savedExperiment.Errors[0].Message)
+		}
+
+	} else {
```
Can we make this code reusable and add it to the utils?

SkySingh04 replied:
Yes, Refactored to DRY!

ispeakc0de commented on pkg/apis/experiment/experiment.go:
```diff
@@ -0,0 +1,348 @@
+	}
+
+	// Query to Run the Chaos Experiment
+	runQuery := `{"query":"mutation{ \n runChaosExperiment(experimentID:  \"` + requestData.ID + `\", projectID:  \"` + pid + `\"){\n notifyID \n}}"}`
```
move this query to pkg/apis/experiment/query.go file

SkySingh04 replied:
refactored

ispeakc0de commented on pkg/apis/experiment/experiment.go:
```diff
@@ -0,0 +1,348 @@
+	}
+}
+
+func SaveExperiment(pid string, requestData model.SaveChaosExperimentRequest, cred types.Credentials) (SaveExperimentData, error) {
```
most of the code of this function is common to CreateExperiment function. Make it reusbale

SkySingh04 replied:
Done, Added as a util function!

ispeakc0de commented on pkg/apis/experiment/experiment.go:
```diff
@@ -0,0 +1,348 @@
+
+}
+
+func RunExperiment(pid string, eid string, cred types.Credentials) (RunExperimentResponse, error) {
```
same as above

SkySingh04 replied:
Done, Added as a util function!

ispeakc0de commented on pkg/apis/experiment/experiment_test.go:
```diff
@@ -0,0 +1,232 @@
+	}
+}
+
+func TestGetExperimentRunsList(t *testing.T) {
```
can we add unit test in Table-Driven Tests format?

```
func TestGetExperimentRunsList(t *testing.T) {
	tests := []struct {
		name     string
		isErr    bool
		execFunc func(clients kubernetes.Interface)
	}{
		{
			name:  "create configmap",
			isErr: false,
			execFunc: func(clients kubernetes.Interface) {
                         // add unit test preparation steps here if any - it is unique for each test
			},
		},
		{
			name:  "testcase name,
			isErr: false,
			execFunc: func(clients kubernetes.Interface) {
				
			},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			fakeK8sClient := fake.NewSimpleClientset([]runtime.Object{}...)
			tt.execFunc(fakeK8sClient)
			c := NewClass(&types.GlobalAssets{
				KubeClient: fakeK8sClient,
				Ctx:        context.Background(),
			})
			err := c.GetExperimentRunsList()
			if tt.isErr {
				require.Error(t, err)
				return
			}
			require.NoError(t, err)
		})
	}
}
```

SkySingh04 replied:
I have refactored the tests approach, Haven't ran them on ubuntu yet and verified. Do let me know if this approach looks good!

ispeakc0de commented on pkg/apis/project.go:
```diff
@@ -0,0 +1,189 @@
+	ProjectName string `json:"projectName"`
+}
+
+func CreateProjectRequest(projectName string, cred types.Credentials) (CreateProjectResponse, error) {
```
same as above

SkySingh04 replied:
Refactored to the best of my capability, let me know if there is any other scope for improvement

ispeakc0de commented on pkg/utils/common.go:
```diff
@@ -0,0 +1,122 @@
+	White   = color.New(color.FgWhite)
+)
+
+func Scanner() string {
```
please remove the unused functions

SkySingh04 replied:
all cleaned up!

SkySingh04 committed:
Update client_test.go

Co-authored-by: Shubham Chaudhary <ashubham314@gmail.com>
Signed-off-by: Sky Singh <akashsingh2210670@gmail.com>

SkySingh04 committed:
Add envconfig integration for Litmus client configuration

- Introduced a new `Configuration` struct in `pkg/utils/variables.go` to hold environment variable configurations for the Litmus client, including endpoint, username, and password.
- Updated the `main.go` file to utilize the new configuration struct, replacing hardcoded values with environment variable support using the `envconfig` package.
- Added the `github.com/kelseyhightower/envconfig` dependency in `go.mod` and updated `go.sum`.

This change enhances the flexibility and security of the Litmus client by allowing configuration through environment variables.

Signed-off-by: Sky Singh <akashsingh2210670@gmail.com>

SkySingh04 committed:
Refactor logging in Litmus client to use logrus

- Replaced standard log package with logrus for enhanced logging capabilities across the application.
- Introduced a new logger package to centralize logging functions, improving consistency and readability.
- Updated various files including `main.go`, `project.go`, and `experiment.go` to utilize the new logging methods.
- Added `github.com/sirupsen/logrus` as a dependency in `go.mod` and updated `go.sum`.

This change enhances the logging functionality, providing better insights into application behavior and errors.

Signed-off-by: Sky Singh <akashsingh2210670@gmail.com>

SkySingh04 committed:
Enhance logging in client tests and main application

- Updated `client_test.go` to log project creation details using the new logger package.
- Refactored error handling in `experiment_test.go` to utilize structured logging for API call errors.
- Removed unused `os` import and a helper function from `main.go` to streamline the code.
- Added new dependencies in `go.mod` and updated `go.sum` accordingly.

These changes improve the clarity and consistency of logging across the application, aiding in debugging and monitoring.

Signed-off-by: Sky Singh <akashsingh2210670@gmail.com>

SkySingh04 committed:
Update copyright year in multiple files to 2025

- Changed the copyright year from 2021 to 2025 in `auth.go`, `project.go`, `request.go`, `experiment.go`, `api_types.go`, and `common.go`.
- This update reflects the current year and ensures proper attribution to the LitmusChaos Authors.

Signed-off-by: Sky Singh <akashsingh2210670@gmail.com>

SkySingh04 committed:
Refactor API request construction in auth and project files

- Updated the construction of API request URLs in `auth.go` and `project.go` to use `fmt.Sprintf` for better readability and maintainability.
- This change enhances the clarity of endpoint definitions and ensures consistent formatting across the codebase.

Signed-off-by: Sky Singh <akashsingh2210670@gmail.com>

SkySingh04 committed:
Refactor experiment API functions to utilize utility methods

- Replaced direct HTTP request handling in `experiment.go` with the new `SendGraphQLRequest` utility for improved readability and error handling.
- Streamlined the process of saving, running, and fetching experiments by consolidating common logic into reusable functions.
- Enhanced error logging consistency across API calls, improving maintainability and debugging capabilities.

This refactor simplifies the codebase and promotes the use of utility functions for GraphQL requests.

Signed-off-by: Sky Singh <akashsingh2210670@gmail.com>

SkySingh04 committed:
Removed unused functions from `common.go`

Signed-off-by: Sky Singh <akashsingh2210670@gmail.com>

SkySingh04 committed:
Enhance test suite with environment variable support and refactor test cases

- Introduced environment variable configuration for test settings in `client_test.go` and `experiment_test.go`, allowing for dynamic test configurations.
- Refactored test cases to utilize a table-driven approach, improving readability and maintainability.
- Updated `README.md` to document the new testing configuration and instructions for running tests with custom settings.
- Added a new test file for project APIs in `pkg/apis/project_test.go`.

These changes improve the flexibility of the test suite and enhance the documentation for developers.

Signed-off-by: Sky Singh <akashsingh2210670@gmail.com>

SkySingh04 committed:
Refactor experiment API to use structured GraphQL queries

- Replaced inline GraphQL mutation strings in `experiment.go` with a structured query constant `RunExperimentQuery` in `query.go`, enhancing code readability and maintainability.
- Updated the `CreateExperiment` and `RunExperiment` functions to utilize the new structured query format, improving consistency in API request handling.

These changes streamline the process of running chaos experiments and promote the use of reusable query structures.

Signed-off-by: Sky Singh <akashsingh2210670@gmail.com>

SkySingh04 committed:
Refactor project API error handling and remove unused test file

- Removed the `project_test.go` file as it was no longer needed.
- Updated error handling in `CreateProjectRequest`, `ListProject`, and `GetProjectDetails` functions to utilize a common `ErrorResponse` struct for improved consistency and readability.
- Streamlined HTTP request handling by replacing direct response body reading with a utility function, enhancing maintainability.

These changes improve the clarity and structure of the project API codebase.

Signed-off-by: Sky Singh <akashsingh2210670@gmail.com>

SkySingh04 committed:
Refactor experiment API to improve request structuring and GraphQL query definitions

- Updated request structs in `CreateExperiment`, `RunExperiment`, `SaveExperiment`, `GetExperimentList`, and `GetExperimentRunsList` functions to use more descriptive field names with JSON tags for better clarity.
- Modified the `RunExperimentQuery` to specify `ID!` for `projectID`, enhancing type safety in GraphQL queries.
- Added an optional `ExperimentRunID` field to the delete operation struct for improved functionality.

These changes enhance the readability and maintainability of the experiment API codebase.

Signed-off-by: Sky Singh <akashsingh2210670@gmail.com>
